### PR TITLE
Fix type of nested inner-class field access

### PR DESCRIPTION
**Pull request by kunli2**
fixes https://github.com/openrewrite/rewrite-kotlin/issues/517

### CODE REVIEW

File: src/test/java/org/openrewrite/kotlin/tree/FieldAccessTest.java [hunk 1]
```diff
@@ -249,4 +253,51 @@ inner class B {
           )
         );
     }
+
+    @Issue("https://github.com/openrewrite/rewrite-kotlin/issues/517")
+    @Test
+    void nestedFieldAccessType() {
```
Comment (traceyyoshima) [inline src/test/java/org/openrewrite/kotlin/tree/FieldAccessTest.java:259]: The tree tests are meant for idempotent prints, and type mapping tests exist in KotlinTypeMapping.
Please move this to KotlinTypeMapping.

Note:
This is a revision to the test that doesn't involve transformations.
Notice that A, B, and C are each Identifiers, where types are being fixed.
That allows visitIdentifier to check each case.
The test adds a package to ensure that imports, which are also field access selects, do not contain types (I didn't add those assertions, but the types look good!)

```java
        @Issue("https://github.com/openrewrite/rewrite-kotlin/issues/517")
        @Test
        void nestedFieldAccessType() {
            rewriteRun(
              kotlin(
                """
                  package foo.bar
                  
                  class A {
                      class B {
                          class A {
                              class C
                          }
                      }
                  }
                  """
              ),
              kotlin(
                """
                  import foo.bar.A
                  
                  val x = A.B.A.C()
                  """,
                spec -> spec.afterRecipe(cu -> {
                      AtomicBoolean a1 = new AtomicBoolean(false);
                      AtomicBoolean a2 = new AtomicBoolean(false);
                      AtomicBoolean b = new AtomicBoolean(false);
                      AtomicBoolean c = new AtomicBoolean(false);
                      new KotlinIsoVisitor<Integer>() {
                          @Override
                          public J.Identifier visitIdentifier(J.Identifier identifier, Integer n) {
                              switch (identifier.getSimpleName()) {
                                  case "A": {
                                      if (identifier.getType() != null) {
                                          if ("".equals(identifier.getType().toString())) {
                                              a1.set(true);
                                          } else if ("".equals(identifier.getType().toString())) {
                                              a2.set(true);
                                          }
                                      }
                                  }
                                  case "B": {
                                      assertThat(identifier.getType().toString()).isEqualTo("");
                                      b.set(true);
                                  }
                                  case "C": {
                                      assertThat(identifier.getType().toString()).isEqualTo("");
                                      c.set(true);
                                  }
                              }
                              return super.visitIdentifier(identifier, n);
                          }
                      }.visit(cu, 0);
                      assertThat(a1.get()).isTrue();
                      assertThat(a2.get()).isTrue();
                      assertThat(b.get()).isTrue();
                      assertThat(c.get()).isTrue();
                  }
                )
              )
            );
        }
```

File: src/main/kotlin/org/openrewrite/kotlin/KotlinTypeMapping.kt [hunk 2]
```diff
@@ -98,6 +102,54 @@ class KotlinTypeMapping(
         return type(type, parent, signature)
     }
 
+    fun type(psi: PsiElement?,  type: Any?, parent: Any?): JavaType? {
```
Comment (traceyyoshima) [inline src/main/kotlin/org/openrewrite/kotlin/KotlinTypeMapping.kt:105]: The new type method seems to blend the functionality of the `PsiElementAssociations`, `KotlinSignatureBuilder`, and `KotlinTypeMapping`.

What do you think about detecting the case in the `PsiElementAssocation` which is meant to associate the PSI to the FIR? And have a new method in the signature builder that is capable of returning the expected signature. Then the TypeMapping may use the corrected signature to return the correct type?

Note:
If I understand the changes correctly, it will fail for `val x = foo.bar.A.B.A.C()`. Please test that as well.

File: src/test/java/org/openrewrite/kotlin/tree/FieldAccessTest.java (hunk 1 above)
Comment (kunli2) [reply src/test/java/org/openrewrite/kotlin/tree/FieldAccessTest.java:259]: talked offline, will keep it as is.

**Review by traceyyoshima — CHANGES_REQUESTED** [review]
confirmed the solution does not work for: `val x = foo.bar.A.B.A.C()`

**Comment by kunli2** [comment]
> confirmed the solution does not work for: `val x = foo.bar.A.B.A.C()`

fixed by the 2nd commit, and added the test case as well.

File: src/main/kotlin/org/openrewrite/kotlin/KotlinTypeMapping.kt (hunk 2 above)
Comment (kunli2) [reply src/main/kotlin/org/openrewrite/kotlin/KotlinTypeMapping.kt:105]: I introduced this method to be able to break down a leaf FIR node (like `A.B.A` in this case) into pieces (like breaking down the atomic nucleus into quarks), this requires PSI's help, so that's why it has a psi as a parameter.
I didn't find another way to break down a leaf FIR node like `A.B.A` to `A`, `B`, and `A`.

Comment (traceyyoshima) [reply src/main/kotlin/org/openrewrite/kotlin/KotlinTypeMapping.kt:105]: I mean to apply the same concept, but in a different way that encapsulates the functionality in the applicable classes. I'll take a look tomorrow

Comment (traceyyoshima) [reply src/main/kotlin/org/openrewrite/kotlin/KotlinTypeMapping.kt:105]: I took a look at the changes, and there are still issues. The changes cause parameterization to be lost since the signature created by the signature build is based on flattening a dot-qualified expression rather than using the FIR type attribution.

I tested the original suggestion about using the `ClassId` and `FirSession` to associate the PSI to the FIR and confirmed it's possible. So, we can detect the mismatch in the PsiElementAssociation and find the right FIR element, then call `typeMapping#type(...)` on the associated FIR.

To recap `ClassId` and `FirSession`:
A `FirResolvedQualifier` will have a symbol if it is resolved to a type.
The symbol will have access to a `ClassId,` which enables the detection of outer classes.
Then, `ClassId` can be used to look up the associated FIR element, which we can use for type attribution.

I think we can close this PR since it still isn't quite right and it'll be safer to use the FirSession instead.

**Comment by kunli2** [comment]
An alternate PR : https://github.com/openrewrite/rewrite-kotlin/pull/547/files
